[V1] Multiprocessing Tensor Parallel Support for v1

tlrmchlsmth opened the pull request:
Implementation of tensor parallel support for V1.

Some differences from V0:
* The executor creates N worker processes rather than N-1 processes like in V0
* All processes run prepare_inputs.
* All processes run the sampler. This means that all workers need the logits so the logits_processor has an AllGather instead of a Gather operation.
* The executor broadcasts the scheduler output to the workers and one of them sends back the model runner output -- both of these IPCs use shared memory message queues.
* The workers sit in a very tight model execution loop where they only handle model execution and process termination.

Some benchmarks are below, using python 3.12. I'm running Llama3-8B on 2 A100s, which will really make any overheads shine.

```
===================
main
===================

python3 benchmarks/benchmark_latency.py --model meta-llama/Llama-3.1-8B-Instruct --tensor-parallel-size 2 \
   --num-iters-warmup 5 --num-iters 20  --batch-size 8 --input-len 128 --output-len 256

Avg latency: 2.301660185540095 seconds
10% percentile latency: 2.2838743404485284 seconds
25% percentile latency: 2.2875737231224775 seconds
50% percentile latency: 2.291928739286959 seconds
75% percentile latency: 2.2938467713538557 seconds
90% percentile latency: 2.3024177022278307 seconds
99% percentile latency: 2.459673847686499 seconds

===================
This PR
===================

VLLM_USE_V1=1 python3 benchmarks/benchmark_latency.py --model meta-llama/Llama-3.1-8B-Instruct --tensor-parallel-size 2 \
   --num-iters-warmup 5 --num-iters 20  --batch-size 8 --input-len 128 --output-len 256

Avg latency: 2.293809377681464 seconds
10% percentile latency: 2.288416426535696 seconds
25% percentile latency: 2.2926345141604543 seconds
50% percentile latency: 2.293652622960508 seconds
75% percentile latency: 2.295131590683013 seconds
90% percentile latency: 2.299927053321153 seconds
99% percentile latency: 2.300706419609487 seconds
```

WoosukKwon commented:
@tlrmchlsmth Thanks for the great work! Let me know when it is ready for review.

njhill commented:
@tlrmchlsmth this looks like a great start and is in the right direction imo. I have been thinking a lot though about how to streamline some things in V1... going to dump some of those ideas here and we can discuss more next week.
- I think we can better unify the single vs multi-GPU logic/abstractions. And actually I think @WoosukKwon's hope is to keep the worker in a separate process even for single GPU, so that we always have n>=1 worker processes.
- We currently have (imo) quite a convoluted layering w.r.t. distributed communication. Initially the idea was to separate control and data planes, but for performance we ended up with this situation where the control plane is an event loop that kicks the workers into a separate loop when executing requests ... in both cases they are waiting on a queue for broadcast messages and I feel it would be simpler to have a single event loop.
- After broadcasting, our core engine loop can perform other CPU tasks while waiting for the response from the rank 0 worker - for example serialization/deserialization of messages to/from the "front-end" process.
- In fact it should be fairly simple (unless I'm missing something) to decouple the data flows to/from the workers and make the scheduling async, so that they always have the schedule for the next step waiting when they finish the current one. We can update and broadcast the schedule with the latest added/removed requests for the (n+2)th step soon after receiving the sampler output for the nth step (i.e. while the (n+1)th step is underway).
- We should use msgpack/msgspec instead of pickle for all these IPC messages.
- I think the current MessageQueue impl (using shm and zmq) would need some adjustments to support the above, such as not using pickle. Also I don't think we would want to use shm for receiving messages in the central proc since it involves spinning. We could use it in the workers to receive the broadcasts but if we are doing async scheduling then it might actually be better to use zmq for a similar reason - and then it will be easier to overlap the I/O and serialization/deserialization with the forward pass by running it in a separate thread.

Mostly orthogonal to the above, I also feel we should rethink the executor abstraction / class hierarchy to better isolate the accelerator-specific aspects. There's a significant amount of duplicated logic right now across different executors (and even workers) and I think we could unify/consolidate a lot of that.

russellb commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -21,6 +23,13 @@
 
 logger = init_logger(__name__)
 
+# We prefer to use os.sched_yield as it results in tighter polling loops,
+# measured to be around 3e-7 seconds. However on python < 3.11.1,
+# os.sched_yield() does not release the GIL, so we fall back to time.sleep(0)
+USE_SCHED_YIELD = False
+if sys.version_info[:3] >= (3, 11, 1):
+    USE_SCHED_YIELD = True
```
You could keep the logic about which method to use here, as well with something like this ...

```suggestion
# We prefer to use os.sched_yield as it results in tighter polling loops,
# measured to be around 3e-7 seconds. However on python < 3.11.1,
# os.sched_yield() does not release the GIL, so we fall back to time.sleep(0)
def sched_yield():
    if sys.version_info[:3] >= (3, 11, 1):
        os.sched_yield()
    else:
        time.sleep(0)
```

njhill replied:
Agree with having a separate function but we might still want to keep the constant to avoid slicing/parsing every time.

nit simplification:
```python
USE_SCHED_YIELD = sys.version_info[:3] >= (3, 11, 1)
```

Also I think the fix was backported to 3.10, we could possibly check for that too.

tlrmchlsmth replied:
Nice, you're right - needs 3.11.1 or 3.10.8
https://docs.python.org/3.11/whatsnew/changelog.html
https://docs.python.org/3.10/whatsnew/changelog.html

russellb commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -1,12 +1,14 @@
 from typing import List, Optional
 from unittest.mock import patch
 
+import msgspec
```
curious if you could say more about what motivated this serialization format change?

njhill replied:
It's much faster

tlrmchlsmth committed:
initial v1 tp support

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
V1 TP with zmq-based boostrapping

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented:
>I think we can better unify the single vs multi-GPU logic/abstractions. And actually I think @WoosukKwon's hope is to keep the worker in a separate process even for single GPU, so that we always have n>=1 worker processes.

Do folks think I should go ahead with this in the current PR? I know @youkaichao had some concerns about ease of debugging if we always put the worker in a separate process. 

Not sure I see a benefit for this implementation, beyond consolidating the implementation and reducing code size. For asynchronous scheduling, it makes a whole lot of sense.

tlrmchlsmth committed:
improve check for USE_SCHED_YIELD

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

tlrmchlsmth committed:
fixup

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

njhill commented:
> > I think we can better unify the single vs multi-GPU logic/abstractions. And actually I think @WoosukKwon's hope is to keep the worker in a separate process even for single GPU, so that we always have n>=1 worker processes.
> 
> Do folks think I should go ahead with this in the current PR? I know @youkaichao had some concerns about ease of debugging if we always put the worker in a separate process.
> 
> Not sure I see a benefit for this implementation, beyond consolidating the implementation and reducing code size. For asynchronous scheduling, it makes a whole lot of sense.

@tlrmchlsmth I think it's fine to defer it.

njhill commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -21,6 +23,12 @@
+# We prefer to use os.sched_yield as it results in tighter polling loops,
+# measured to be around 3e-7 seconds. However on earlier versions of Python
+# os.sched_yield() does not release the GIL, so we fall back to time.sleep(0)
+USE_SCHED_YIELD = ((sys.version_info[:3] >= (3, 11, 1))
+                   or (sys.version_info[:3] >= (3, 10, 8)))
```
In theory this would match 3.11.0 otherwise I think...
```suggestion
USE_SCHED_YIELD = ((sys.version_info[:3] >= (3, 11, 1))
                   or (sys.version_info[:2] == (3, 10) and sys.version_info[2] >= 8))
```

tlrmchlsmth replied:
thanks for the catch!

tlrmchlsmth committed:
workers must be daemonic

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
We can now terminate properly

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
fixes from merge

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Fixup termination

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented:
Still a few open issues but marking it ready for review as I work though them

tlrmchlsmth commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +226,278 @@ def profile(self, is_start=True):
+
+    # Main busy loop for Multiprocessing Workers
+    def execute_model_busy_loop(self):
+        with torch.profiler.profile(
+                activities=[
+                    torch.profiler.ProfilerActivity.CPU,
+                    torch.profiler.ProfilerActivity.CUDA,
+                ],
+                schedule=torch.profiler.schedule(
+                    wait=1000,  # Wait 1000 steps so we profile middle iters
+                    warmup=10,  # Warm up the scheduler
+                    active=3,  # Run a small number of steps so it's legible
+                    repeat=1,
+                ),
+                on_trace_ready=torch.profiler.tensorboard_trace_handler(
+                    "./traces/",
+                    worker_name=f"worker_{self.worker.rank}",
+                ),
+                with_stack=True,
```
Reminder to remove debug cruft before landing

tlrmchlsmth committed:
Appease mypy

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Allow shm_broadcast to enqueue by eithr pickle or msgpack

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

bnellnm commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -440,6 +468,57 @@ def dequeue(self):
+            self.remote_socket.send(serialized_obj)
+
+    def dequeue_via_msgpack(self, obj_type):
+        """Enqueue obj using msgpack serialization"""
```
nit: type `Enqueue` -> `Dequeue`

njhill reviewed (commented):
Thanks @tlrmchlsmth this is looking good.

I have a few more suggestions:
- I don't think we need a separate `start_workers` method, instead it may be cleaner to rename the `initialize_cache` method to `initialize` and assume that if/when the executor returns from this successfully then it's ready
- This would mean that `BEGIN_MODEL_EXECUTION` and the `workers_in_busy_loop` flag can be removed, but `INIT_CACHE` would have to return a success/failure message
- We could even consider having `initialize` at the executor level encompass both `determine_num_blocks` and `initialize_cache` (push down the num blocks override logic) .. but this could be a follow-on thing
- I think it's fine to keep things consistent, but if it's simpler we could use pickle for all the init stuff - IPC overhead isn't a concern here
- Slightly orthogonal but we should rename `gpu_executor` and `multiproc_gpu_executor` to something like `sameproc_executor` and `multiproc_executor` - they can be platform agnostic (I am making a similar simplification to v0)

I don't think this should block merging, but there's also quite a bit of error handling still to be added, in particular individual workers failing at various stages.

I still wonder whether we should allow for splitting execute_model into separate send/receive, to allow cleaner coordination of other work while waiting for the output in the engine. The way we have this in a completely independent thread currently isn't ideal because we really want it to run only while the engine loop is waiting. But I think this is also something that can be refactored as follow-on work if needed.

tlrmchlsmth committed:
Switch back to pickle for shm_broadcast serialization

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented:
>I think it's fine to keep things consistent, but if it's simpler we could use pickle for all the init stuff - IPC overhead isn't a concern here

After discussion with @WoosukKwon last night, I'm going to switch this PR to use `pickle`. The main sticking point is multimodal inputs as `mspack` does not support serializing numpy.ndarray (we could use `tobytes` but this is much slower)

tlrmchlsmth committed:
Finish msgpack -> pickle

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
wrap sched_yield and time.sleep in a fn

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented on vllm/v1/executor/multiproc_gpu_executor.py:
```diff
@@ -0,0 +1,165 @@
+        underlying worker.
+        """
+        # Get the maximum number of blocks that can be allocated on GPU and CPU.
+        num_blocks = self.run_on_workers('determine_num_available_blocks')
+
+        # Since we use a shared centralized controller, we take the minimum
+        # number of blocks across all workers to make sure all the memory
+        # operators can be applied to all workers.
+        num_gpu_blocks = min(b[0] for b in num_blocks)
+        num_cpu_blocks = min(b[1] for b in num_blocks)
```
no, because we get list of tuples rather than a tuple of lists here

njhill replied:
ah of course, my bad

tlrmchlsmth commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -329,7 +349,10 @@ def acquire_write(self):
 
                     # Release the processor to other threads
-                    os.sched_yield()
+                    if USE_SCHED_YIELD:
+                        os.sched_yield()
+                    else:
+                        time.sleep(1e-5)
```
Fixed, yeah pulling into a function helps here

tlrmchlsmth committed:
Review comments

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Rename executors to uniproc and multiproc

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
more review comments

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
format

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
hacky hacky hacky cleanup

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Fix spawn vs fork issue using approach from #8823

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
skip non-distributed tests in test_basic_correctness to see what happens

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

rkooo567 commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -99,10 +98,21 @@ def from_engine_args(
+
+        return executor_class
 
     def stop_remote_worker_execution_loop(self) -> None:
```
QQ: what is this API for? IIUC, we don't need this API anymore with new scheduler <> worker decoupled design?

tlrmchlsmth replied:
Are you talking about `stop_remote_worker_execution_loop` or `_get_executor_cls`?

For `stop_remote_worker_execution_loop` I am not sure -- honestly I stumbled across this while I was trying to debug some process termination bugs.

njhill replied:
@rkooo567 yes you're right, this will be no longer needed/used since we'll have a single event loop in the workers, we should remove it!

tlrmchlsmth replied:
I'll take it out 👍

tlrmchlsmth committed:
fix async_llm

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
format

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

bnellnm commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -99,10 +98,20 @@ def from_engine_args(
     @classmethod
     def _get_executor_cls(cls, vllm_config: VllmConfig):
-        return GPUExecutor
+        distributed_executor_backend = (
+            vllm_config.parallel_config.distributed_executor_backend)
+        if distributed_executor_backend == "mp":
+            from vllm.v1.executor.multiproc_executor import MultiprocExecutor
+            executor_class = MultiprocExecutor
+        else:
+            assert (distributed_executor_backend is None)
+            from vllm.v1.executor.uniproc_executor import UniprocExecutor
+            executor_class = UniprocExecutor
+
+        return executor_class
```
Is it worth pulling this code out to a utility function somewhere?

bnellnm commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,168 @@
+
+    def shutdown(self):
+        """Properly shut down the executor and its workers"""
+        if (hasattr(self, 'scheduler_output_mq')
```
Is this hasattr check necessary for partially constructed objects?

tlrmchlsmth committed:
Fixes for testing

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Abstract executor class for typing

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
remove enforce_eager, format

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
remove stop_remote_worker_execution_loop

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Remove profiling

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
ExecutorMsg -> WorkerExecRequest

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

WoosukKwon commented on vllm/v1/executor/uniproc_executor.py:
QQ: Why do we have two executors?

tlrmchlsmth replied:
The UniprocExecutor avoids the overheads of serialization and IPC, and I think it will make debugging easier.

We can take it out and only have the MultiprocExecutor -- I think it make sense to do this if we either move to async scheduling or if we otherwise break up `execute_model` so that the engine core process can do other things during the forward pass.

WoosukKwon replied:
Do you have performance benchmarks by any chance?

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

tlrmchlsmth committed:
ensure_termination

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Move ensure_termination to executor to avoid futures

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

SageMoore commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +229,311 @@ def profile(self, is_start=True):
+                raise ValueError(f"Unknown RequestType: {response_type}")
+            return pickle.loads(response_data)
+
+    def initialize(self, num_gpu_blocks: int) -> bool:
```
It looks like we could remove some of the duplicated code by having something like:

`def make_remote_request(request_type: WorkerInitRequestType, msg: Optional[]): `

The above function could handle all of the socket management.

tlrmchlsmth replied:
I think this code is a little too immature for a DRY pass so I'll hold off on this and we can refactor in a future PR

SageMoore reviewed (commented):
Generally looks good. Just minor comments.

tlrmchlsmth commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +229,311 @@ def profile(self, is_start=True):
+        self.worker.initialize()
+        self.worker.load_model()
+
+    @staticmethod
```
This is what @robertgshaw2-neuralmagic did for AsyncLLM and I liked how it organized the code so I copied it

tlrmchlsmth committed:
minor updates

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
call destroy_distributed_environment atexit

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
more graceful shutdown

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Simplify worker termination

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
atexit -> weakref.finalize

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
minor cleanup

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
core client cleanup rework

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
poke CI

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
fix V1 test, temporarily delete some noisy log statements

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
nccl/issues/1234

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Cleanup, _add_prefix

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
fixup noise a bit

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
tweaks

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

tlrmchlsmth committed:
back to atexit

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

robertgshaw2-neuralmagic commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,197 @@
+        scheduler_output,
+    ) -> ModelRunnerOutput:
+        self.worker_request_mq.enqueue(scheduler_output)
+        model_output = self.model_output_mq.dequeue()
```
This is a blocking call right?

tlrmchlsmth replied:
Yes and yes -- check out shm_broadcast.py. Had to sidestep an issue there to make sure `enqueue` and `dequeue` release the GIL

robertgshaw2-neuralmagic replied:
okay I see now how the `acquire_read` and `acquire_write` are used in this pathway. Nice find on the time.sleep()!

robertgshaw2-redhat reviewed (approved):
Nice work. Left a few minor comments which can be addressed in a follow up

robertgshaw2-neuralmagic commented on vllm/v1/engine/core.py:
```diff
@@ -240,17 +221,28 @@ def make_engine_core_process(
             logger.exception(e)
             raise e
 
+        finally:
```
tiny nit
- If `EngineCoreProc()` fails, then `engine_core` will not exist in `finally`
- I think you can get around this by setting `engine_core = None` outside the try block and then checking if it is None in the finally block before calling shutdown

robertgshaw2-neuralmagic commented on vllm/v1/engine/core_client.py:
```diff
@@ -156,15 +161,16 @@ def __init__(
     def shutdown(self):
         # Send shutdown signal to background process.
         self.should_shutdown = True
+        self.proc.join(5)
```
tiny nit:
- it is possible that `self.proc` does not exist here
- Do we need to call `.join(5)` twice?

robertgshaw2-neuralmagic commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,197 @@
+        self.model_output_mq = None
+        self.worker_request_mq = None
+
+    def check_health(self) -> None:
```
In a follow up, this should ping the workers to see if they are alive

robertgshaw2-neuralmagic commented on vllm/v1/outputs.py:
```diff
@@ -29,7 +31,7 @@ class ModelRunnerOutput:
 
     # [num_reqs]
-    sampled_token_ids_cpu: torch.Tensor
+    sampled_token_ids_cpu: List[int]
```
tiny nit -> this does not need the training `_cpu` suffix. The convention in V1 is that `_cpu` is used to torch cpu objects

robertgshaw2-neuralmagic commented on vllm/v1/sample/sampler.py:
```diff
@@ -38,7 +38,7 @@ def forward(
 
         sampler_output = SamplerOutput(
-            sampled_token_ids=sampled,
+            sampled_token_ids=sampled.tolist(),
```
is this where the `CPU<>GPU` sync happens?

robertgshaw2-neuralmagic commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +230,279 @@ def profile(self, is_start=True):
+                else:
+                    raise ValueError(f"Unknown RequestType: {request_type}")
+
+    def execute_model_busy_loop(self):
```
I would recommend wrapping this in a try..catch and logging any exceptions. Since this is going to run in the background process, it will help significantly to debug issues.

robertgshaw2-neuralmagic replied:
@tlrmchlsmth - this is something that has caused pain with the `MQLLMEngine` design.

robertgshaw2-neuralmagic replied:
Nevermind, i see that the caller of this function wraps with a try...catch

robertgshaw2-neuralmagic commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +230,279 @@ def profile(self, is_start=True):
+            logger.exception(e)
+            raise
+
+        finally:
```
if `worker = WorkerProc` fails, then calling worker.shutdown will raise an exception

```python
def raise_error():
    raise Exception
    
try:
     a = raise_error()
finally:
     del a
```

gives:

```bash
>>> NameError: name 'a' is not defined
```

WoosukKwon commented:
@tlrmchlsmth I found that the worker processes are not properly cleaned up while the main process was terminated normally without any error message.

tlrmchlsmth commented:
@WoosukKwon thanks for trying it out, can you give repro steps for that?

WoosukKwon commented:
@tlrmchlsmth I just tried `$ VLLM_USE_V1=1 VLLM_ENABLE_V1_MULTIPROCESSING=1 python benchmarks/benchmark_latency.py -tp 2` a few times. The bug doesn't always happen, but occasionally it leaves the zombie worker processes.

WoosukKwon commented:
This actually happens pretty often... It almost always happen for `$ VLLM_USE_V1=1 VLLM_ENABLE_V1_MULTIPROCESSING=1 python3 benchmarks/benchmark_latency.py --model meta-llama/Llama-3.1-8B-Instruct --tensor-parallel-size 2 --num-iters-warmup 5 --num-iters 20  --batch-size 8 --input-len 128 --output-len 256 --no-enable-prefix-caching`

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,197 @@
+            model_output_mq_handle, 0)
+
+    def _run_on_workers(self, fn: str, *args) -> List:
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(getattr(type(w), fn), w, *args)
+                for w in self.workers
+            ]
+            result = [f.result() for f in futures]  # Wait for all to complete
```
is it possible to directly send the arguments via the shared memory broadcast? here we are serializing the `*args` for every worker i think.

tlrmchlsmth replied:
This is only used during worker initialization, so I'm not too worried about performance here. I'll add a comment making this clear.

We could use the shared memory broadcast for IPC during initialization, but the way I've designed the PR is that `WorkerProcHandle` is responsible for all the initialization messages, and I'm using ZMQ there because it has fewer gotchas.

tlrmchlsmth replied:
Just to emphasize: The main difference between this implementation and the one in `multiproc_worker_utils.py` is in the latter, the workers have one flexible event loop that's used for all messages. Here, we have two: one flexible event loop used only for initialization (uses ZMQ), and one very rigid very tight even loop only for model execution (I think we'll need to extend this at least to handle check health)

youkaichao replied:
if the flexible event loop is only used for initialization, then we don't need to abstract it to a function?

we should have a general `run_workers` function , it's not just about model execution, but also adding/unloading loras, and maybe other user-defined function.

my design would be to always use the very tight event loop ( shared memory broadcast) for all the function call.

tlrmchlsmth replied:
Added a general `collective_rpc` function. PTAL!

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,197 @@
+                           "VLLM_WORKER_MULTIPROC_METHOD to 'spawn'.")
+            os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
+
+        # Ensure that VLLM_INSTANCE_ID is set, to be inherited by workers
+        os.environ["VLLM_INSTANCE_ID"] = get_vllm_instance_id()
+
+        # Disable torch async compiling which won't work with daemonic processes
+        os.environ["TORCHINDUCTOR_COMPILE_THREADS"] = "1"
+
+        # Configure thread parallelism if OMP_NUM_THREADS isn't set
+        #
+        # Helps to avoid CPU contention. The default of spawning a thread per
+        # core combined with multiprocessing for each GPU can have a negative
+        # impact on performance. The contention is amplified when running in a
+        # container where CPU limits can cause throttling.
+        default_omp_num_threads = 1
+        if "OMP_NUM_THREADS" not in os.environ and (
+                current_parallelism :=
+                torch.get_num_threads()) > default_omp_num_threads:
+            logger.warning(
+                "Reducing Torch parallelism from %d threads to %d to avoid "
+                "unnecessary CPU contention. Set OMP_NUM_THREADS in the "
+                "external environment to tune this value as needed.",
+                current_parallelism, default_omp_num_threads)
+            os.environ["OMP_NUM_THREADS"] = str(default_omp_num_threads)
+            torch.set_num_threads(default_omp_num_threads)
+
+        # workaround for https://github.com/vllm-project/vllm/issues/6103
+        if world_size > 1:
+            maybe_set_triton_cache_manager()
```
I don't really like that we duplicate lots of code here 😢

youkaichao replied:
please use https://github.com/vllm-project/vllm/blob/9743d64e4e04a88174c76553fcbffa33a18c7db5/vllm/worker/worker_base.py#L408 so that we can have a common place to inject process initialization logic.

WoosukKwon replied:
@tlrmchlsmth Can you please refactor this out to reduce the duplication?

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,197 @@
+        self,
+        scheduler_output,
+    ) -> ModelRunnerOutput:
+        self.worker_request_mq.enqueue(scheduler_output)
```
I think all `_run_on_workers` can be implemented this way?

tlrmchlsmth committed:
Clean up process termination

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
robcomments

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
format

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

youkaichao commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +231,290 @@ def profile(self, is_start=True):
+                kwargs["initialization_input_path"],
+                kwargs["initialization_output_path"])
+
+            worker.execute_model_busy_loop()
```
having a busy loop is fine, but please do not only use it for model execution, it will not be extensible, e.g.:

- loading / unloading lora
- RLHF usecase I'm working on, where we need to perform some weight update and then continue execution.

tlrmchlsmth replied:
Yes, I will add other message types. I think this should be done in a subsequent PR

tlrmchlsmth committed:
client now kills workers directly to avoid zombies

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented on tests/basic_correctness/test_basic_correctness.py:
```diff
@@ -143,6 +157,7 @@ def test_models_distributed(
     )
 
 
+@pytest.mark.skip_v1
```
This test fails on V1 but I don't know why. It's not related to this PR as it's not running TP and fails on current main (just enabled it on https://github.com/vllm-project/vllm/pull/10864)

tlrmchlsmth commented on vllm/compilation/backends.py:
```diff
@@ -247,11 +247,6 @@ def __call__(self, graph: fx.GraphModule, example_inputs) -> Callable:
         self.split_gm, self.piecewise_graphs = split_graph(
             graph, self.compilation_configs.splitting_ops)
 
-        from torch._dynamo.utils import lazy_format_graph_code
-        logger.debug("%s", lazy_format_graph_code("before split", self.graph))
-        logger.debug("%s", lazy_format_graph_code("after split",
-                                                  self.split_gm))
```
Going to revert this, but I deleted these because they made the logs very noisy and difficult to read while I was trying to debug an unrelated error. Maybe we can add a `debug_verbose` logging level and use it here?

bnellnm replied:
I'd be in favor of deleting these.  I think we should switch these over to use the same `dump_graph` system used for compilation passes.

tlrmchlsmth replied:
Ok, I'm going to leave them deleted then :)

youkaichao replied:
please leave it unchanged, i will improve it at https://github.com/vllm-project/vllm/pull/10972 .

tlrmchlsmth commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -210,6 +231,290 @@ def profile(self, is_start=True):
+                        logger.exception(e)
+
+                        # Send a failure response
+                        send_socket.send_multipart(
```
I'm not doing this. I also don't have check_health implemented (as @robertgshaw2-neuralmagic pointed out previously) 

How does V0 handle these cases?

tlrmchlsmth committed:
remote rpc

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
use WorkerWrapperBase

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
de-duplicate env setup

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented:
I think I've addressed all existing comments.

One note is that I've incorporated @youkaichao's idea for a general shm_broadcast-based "run on workers" method. It's very flexible, but as a consequence will make it difficult to use `msgpack` for serialization if we choose to that in the future

youkaichao commented:
what are the general requirements for messages to be serialized by `msgpack` ? we can also change `shm_broadcast` implementation to `msgpack` if possible, maybe in a later pr.

tlrmchlsmth commented:
> what are the general requirements for messages to be serialized by `msgpack` ? we can also change `shm_broadcast` implementation to `msgpack` if possible, maybe in a later pr.

In a previous revision of this PR I extended shm_broadcast to support msgpack. It’s not too bad. 

* I had to prepend the encoded message with its length 
* When decoding, you need to know the type of the object that you’re decoding (this conflicts with the flexibility of the collective_rpc implementation)
* numpy ndarrays aren’t supported 😢

youkaichao commented:
> When decoding, you need to know the type of the object that you’re decoding

really? I think a decent serialization scheme should encode this message in the payload. didn't expect msgpack is so inflexible.

youkaichao commented on tests/basic_correctness/test_basic_correctness.py:
```diff
@@ -118,6 +127,11 @@ def test_models_distributed(
         os.environ["VLLM_ATTENTION_BACKEND"] = attention_backend
 
+    # Import VLLM_USE_V1 dynamically to handle patching
+    from vllm.envs import VLLM_USE_V1
```
what does this dynamic patching mean? `envs.VLLM_USE_V1` should read the latest env var value.

tlrmchlsmth replied:
I moved the import here to get the `VLLM_USE_V1` check using when we are using the `run_with_both_engines` pytest fixture during testing.

https://github.com/vllm-project/vllm/blob/bf0e382e16065edebbbb414f7889d31523a569e1/tests/conftest.py#L112-L126

Please LMK if you have a better idea!

youkaichao commented on vllm/platforms/cuda.py:
```diff
@@ -100,6 +100,13 @@ def log_warnings(cls):
+
+        from vllm.envs import VLLM_USE_V1
+        if VLLM_USE_V1:
+            assert parallel_config.worker_cls == "auto"
```
please do not assert `auto` , it will disable future integration.

please place this branch under `if parallel_config.worker_cls == "auto":` , and directly return `"vllm.v1.worker.gpu_worker.Worker"` if you find it is v1.

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,171 @@
+        """
+        Initialize the KV caches and begin the model execution loop of the
+        underlying workers.
+        """
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(w.initialize, num_gpu_blocks)
+                for w in self.workers
+            ]
+            success_vals = [f.result()
+                            for f in futures]  # Wait for all to complete
+
+        if not all(success_vals):
+            raise RuntimeError("Worker initialization failed.")
+
+        self.worker_request_mq.wait_until_ready()
+        for w in self.workers:
+            w.worker_response_mq.wait_until_ready()
+
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        """
+        Determine the number of available KV blocks by invoking the
+        underlying worker.
+        """
+        # Get the maximum number of blocks that can be allocated on GPU and CPU.
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(w.determine_num_available_blocks)
+                for w in self.workers
+            ]
+            num_blocks = [f.result()
+                          for f in futures]  # Wait for all to complete
+
+        # Since we use a shared centralized controller, we take the minimum
+        # number of blocks across all workers to make sure all the memory
+        # operators can be applied to all workers.
+        num_gpu_blocks = min(b[0] for b in num_blocks)
+        num_cpu_blocks = min(b[1] for b in num_blocks)
+
+        return num_gpu_blocks, num_cpu_blocks
```
this logic looks strange to me. i think we should first set up the connection via 

```python
        self.worker_request_mq.wait_until_ready()
        for w in self.workers:
            w.worker_response_mq.wait_until_ready()
```

and then `determine_num_available_blocks` and `initialize` are just normal `collective_rpc` calls.

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,171 @@
+
+        return num_gpu_blocks, num_cpu_blocks
+
+    def collective_rpc(self, fn: str, timeout: Optional[float], *args,
```
nit: `fn` -> `method: str`

youkaichao replied:
we should also add `output_ranks` option, it's not always the first rank returns something, users can request the output from multiple workers.

and since we have `responses`, it should be easy to just return corresponding slices.

youkaichao commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -209,6 +229,307 @@ def profile(self, is_start=True):
+        return
+
+
+@dataclass
+class WorkerInitRequestType:
+    """
+    Request types defined as hex byte strings, so it can be sent over sockets
+    without separate encoding step.
+    """
+    DETERMINE_NUM_BLOCKS = b'\x00'
+    INITIALIZE = b'\x01'  # Initialize cache and begin worker execution
+
+
+@dataclass
+class WorkerInitResponseType:
+    """
+    Request types defined as hex byte strings, so it can be sent over sockets
+    without separate encoding step.
+    """
+    READY = b'\x00'
+    NUM_BLOCKS = b'\x01'
+    INITIALIZE_SUCCESS = b'\x02'
```
we should not need this if we set up the connection early, and all these can be normal `collective_rpc` calls.

youkaichao commented on vllm/v1/executor/abstract.py:
```diff
@@ -0,0 +1,40 @@
+from vllm.v1.outputs import ModelRunnerOutput
+
+
+class Executor(ABC):
```
at a high level, i think `class Executor` should only have:

```python
class Executor:
    def __init__(self, vllm_config: VllmConfig):

    def collective_rpc(self, ...):
```

the engine should only interact with the executor via `collective_rpc`

tlrmchlsmth replied:
Leaving this for now, until we axe `UniprocExecutor`

youkaichao replied:
sounds good to me right now. please keep in mind and refactor later.

tlrmchlsmth committed:
Use collective_rpc for initialization

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
add RPCParams for readability

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
fixup

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp: instance id

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth commented:
Switched initialization over to use `collective_rpc`. Easy now that we've made it general

youkaichao commented on vllm/model_executor/layers/logits_processor.py:
```diff
@@ -42,7 +42,9 @@ def __init__(self,
         self.soft_cap = soft_cap
         # Whether to use gather or all-gather to gather the logits.
-        self.use_gather = not current_platform.is_tpu()
+        # note: we import VLLM_USE_V1 dynamically to handle patching
+        from vllm.envs import VLLM_USE_V1
```
`vllm.envs` has a module level `__getattr__` , and the recommended way is to use `envs.VLLM_USE_V1` . I don't know if `from vllm.envs import` works.

youkaichao commented on vllm/platforms/cuda.py:
```diff
@@ -100,17 +100,30 @@ def log_warnings(cls):
         parallel_config = vllm_config.parallel_config
         scheduler_config = vllm_config.scheduler_config
+
+        from vllm.envs import VLLM_USE_V1
```
same comment as above

youkaichao commented on vllm/v1/executor/abstract.py:
```diff
@@ -0,0 +1,53 @@
+        raise NotImplementedError
+
+    @abstractmethod
+    def collective_rpc(self, rpc: RPCParams, *args, **kwargs) -> [Optional]:
```
after a second thought, I think `output_ranks` are not necessary. we should:

```python
def collective_rpc(self, method: str, timeout: Optional[float]=None, *args, **kwargs) -> List[Any]:
    pass
```

`collective_rpc` should check all workers return correctly, collect all the outputs, and 

```python
    def execute_model(
        self,
        scheduler_output,
    ) -> ModelRunnerOutput:
        return self.collective_rpc(method="execute_model", scheduler_output)[-1]
```

some rpc calls can filter out the objects they want, depending on the semantic.

tlrmchlsmth replied:
What advantage do you see over the current design? If we do this then each worker process will serialize and send the model output regardless of if it's needed. And the current message queue implementation means that the executor will have to deserialize all of them

youkaichao replied:
> If we do this then each worker process will serialize and send the model output regardless of if it's needed.

you don't need to send the model output. for non-zero rank workers, even if it has the model output, it should send a `None` object as a placeholder.

in the worker side, it should be:

```python
def execute_model(self, scheduler_output) -> Optional[ModelRunnerOutput]:
    output = self.model_runner(scheduler_output)
    if self.rank != 0:
        return None
    return output
```

The idea is that all workers need to send **something**, to indicate that it executes successfully. otherwise, even if you get the output from rank 0, you don't know if other workers have errors (in which case you should not return the output, but should raise errors).

youkaichao replied:
why is this resolved?

tlrmchlsmth replied:
not sure but definitely should be resolved now

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,168 @@
+
+        # Initialize worker and set up message queues for SchedulerOutputs
+        # and ModelRunnerOutputs
+        self.worker_request_mq = MessageQueue(self.world_size, self.world_size)
```
nit: rename to `rpc_input_mq` or something like that? `request` is too specialized, especially when people integrate general `collective_rpc` functions.

tlrmchlsmth replied:
how about `rpc_broadcast_mq`? IMO `input` is tricky because we deal with both sides of the message queue, and it's helpful to be able to use one name consistently in the worker process and the engine core process

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,168 @@
+            responses = [None] * self.world_size
+            for w in self.workers:
+                status, result = w.worker_response_mq.dequeue(
+                    timeout=rpc.timeout)
```
nit: `timeout` should have a common starting time.

tlrmchlsmth replied:
oh yeah, this is a bug - nice catch!

tlrmchlsmth replied:
Switched the shm MessageQueue interface to take a deadline instead of a timeout

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,168 @@
+        for w in self.workers:
+            w.worker_response_mq.wait_until_ready()
+
+    def initialize(self, num_gpu_blocks: int) -> None:
+        """
+        Initialize the KV caches and begin the model execution loop of the
+        underlying workers.
+        """
+        self.collective_rpc(
+            RPCParams(method="initialize_cache", output_ranks={}),
+            num_gpu_blocks)
+        self.collective_rpc(
+            RPCParams(method="compile_or_warm_up_model", output_ranks={}))
+
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        """
+        Determine the number of available KV blocks by invoking the
+        underlying worker.
+        """
+        num_blocks = self.collective_rpc(
+            RPCParams(method="determine_num_available_blocks",
+                      output_ranks=set(range(self.world_size))))
+
+        # Since we use a shared centralized controller, we take the minimum
+        # number of blocks across all workers to make sure all the memory
+        # operators can be applied to all workers.
+        num_gpu_blocks = min(b[0] for b in num_blocks)
+        num_cpu_blocks = min(b[1] for b in num_blocks)
+
+        return num_gpu_blocks, num_cpu_blocks
+
```
I want to merge them into one single `ready_for_serving`, so that we can easily profile the starting time of vllm. but this is out of the scope of this pr.

tlrmchlsmth replied:
I'm in favor of this. Will generally clean up the initialization process as well

youkaichao commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -209,6 +229,193 @@ def profile(self, is_start=True):
+        return
+
+
+@dataclass
+class WorkerProcHandle:
+    proc: BaseProcess
+    rank: int
+    ready_path: str
+    worker_response_mq: MessageQueue  # The worker process writes to this MQ
+
+
+class WorkerProc:
+    """Wrapper that runs one Worker in a separate process."""
+
+    READY_STR = "READY"
```
we should move both classes into `multiproc_executor.py` , or follow the v0 design, to have a `multiproc_worker_utils.py` .

the motivation for `WorkerWrapperBase`, is that we want to do something even before importing the `worker_cls` . if we put both classes here, before `WorkerWrapperBase` is executed, the `worker_cls` is already imported.

tlrmchlsmth replied:
Moved them. TIL about the import issue

youkaichao commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -209,6 +229,193 @@ def profile(self, is_start=True):
+            input_shm_handle, self.worker.rank)
+
+        # Initializes a message queue for sending the model output
+        self.worker_response_mq = MessageQueue(1, 1)
```
we discussed in the slack that `MessageQueue` with a single reader might have minimal benefit and we can just use a `XSUB` from the executor side connected with `n` `PUB` from all workers. will that be implemented in this pr?

tlrmchlsmth replied:
I think we should land the current PR using shared memory message queues, and then I think there are a few different options we can explore

youkaichao reviewed (commented):
thanks for the great efforts! the pr looks good to me in general, I left several new comments, but they should be easy to address.

I do want to discuss one more thing about the initialization process tomorrow.

tlrmchlsmth committed:
review comments

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

tlrmchlsmth committed:
profile

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
move vllm envs import to work with run_with_both_engines

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Merge branch 'main' into tms/v1_tp

youkaichao commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -406,24 +430,26 @@ def acquire_read(self):
             self.remote_socket.send(serialized_obj)
 
-    def dequeue(self):
+    def dequeue(self, deadline: Optional[float] = None):
```
why do you name it `deadline` ? I think `timeout` should be a common name.

youkaichao commented on vllm/model_executor/layers/logits_processor.py:
```diff
@@ -42,7 +42,13 @@ def __init__(self,
+
+        # Import vllm.envs here in order to to work with the
+        # run_with_both_engines pytest fixture
+        import vllm.envs as envs
```
you can put it at the top. it does not matter.

everytime you use `envs.VLLM_USE_V1` , it is a function call and will read the latest env var.

tlrmchlsmth replied:
It looks like the first time we use `envs.VLLM_USE_V1`, it puts the value of the environment variable in a dict and then doesn't query the OS on subsequent uses.

is that right?

youkaichao replied:
no, it is a module level access function. whenever you use `envs.VLLM_USE_V1` , it runs the function again.

youkaichao commented on vllm/platforms/cuda.py:
```diff
@@ -110,17 +110,32 @@ def log_warnings(cls):
+
+        # Import vllm.envs here in order to to work with the
+        # run_with_both_engines pytest fixture
+        import vllm.envs as envs
```
same as above.

youkaichao commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -0,0 +1,360 @@
+        """Execute a collective remote procedure call on workers"""
+        self.rpc_broadcast_mq.enqueue((rpc.method, args, kwargs))
+
+        response_deadline = (time.monotonic() +
```
you can record the starting time here, and shrink the `timeout` parameter for `dequeue`

youkaichao reviewed (approved):
thanks for the great work! I don't have major concern now, there are some followup work, but we don't necessarily need to do it in this pr.

before merge, please fix the nit comments about `vllm.envs` , and change the signature of `collective_rpc` to `collective_rpc(self, method, timeout, *args, **kwargs)`, thanks.

tlrmchlsmth committed:
review comments.

- update collective_rpc interface
- change run_on_both_enginges pytorch fixture

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
collective rpc function signature sanity

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

youkaichao commented:
LGTM to merge now.

youkaichao commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -383,16 +405,20 @@ def acquire_read(self):
-                        logger.warning(
-                            "No available block found in %s second. ",
-                            VLLM_RINGBUFFER_WARNING_INTERVAL)
+                        logger.debug("No available block found in %s second. ",
```
it seems we don't log this message at all by default now?

tlrmchlsmth replied:
@youkaichao any idea how to suppress this message when vLLM is idle? Right now on current main if we're not serving any requests, we will tell users that vLLM is probably hanging every 60 seconds 😆
